stoix/systems/q_learning/ff_noisy_dqn.py

```diff
@@ -0,0 +1,569 @@
+    timesteps = jax.tree_util.tree_map(reshape_states, timesteps)
+
+    # Load model from checkpoint if specified.
+    if config.logger.checkpointing.load_model:
+        loaded_checkpoint = Checkpointer(
+            model_name=config.system.system_name,
+            **config.logger.checkpointing.load_args,  # Other checkpoint args
+        )
+        # Restore the learner state from the checkpoint
+        restored_params, _ = loaded_checkpoint.restore_params(TParams=OnlineAndTarget)
+        # Update the params
+        params = restored_params
+
+    # Define params to be replicated across devices and batches.
+    key, step_key, warmup_key = jax.random.split(key, num=3)
+    step_keys = jax.random.split(step_key, n_devices * config.arch.update_batch_size)
+    warmup_keys = jax.random.split(warmup_key, n_devices * config.arch.update_batch_size)
+
+    def reshape_keys(x):
+        return x.reshape((n_devices, config.arch.update_batch_size) + x.shape[1:])
+
+    step_keys = reshape_keys(jnp.stack(step_keys))
+    warmup_keys = reshape_keys(jnp.stack(warmup_keys))
+
+    replicate_learner = (params, opt_states, buffer_states)
+
+    # Duplicate learner for update_batch_size.
+    def broadcast(x):
+        return jnp.broadcast_to(x, (config.arch.update_batch_size,) + x.shape)
+
+    replicate_learner = jax.tree_util.tree_map(broadcast, replicate_learner)
+
+    # Duplicate learner across devices.
+    replicate_learner = flax.jax_utils.replicate(replicate_learner, devices=jax.devices())
+
+    # Initialise learner state.
+    params, opt_states, buffer_states = replicate_learner
+    # Warmup the buffer.
+    env_states, timesteps, keys, buffer_states = warmup(
+        env_states, timesteps, buffer_states, warmup_keys
+    )
+    init_learner_state = OffPolicyLearnerState(
+        params, opt_states, buffer_states, step_keys, env_states, timesteps
+    )
+
+    return learn, eval_q_network, init_learner_state
+
+
+def run_experiment(_config: DictConfig) -> float:
+    """Runs experiment."""
+    config = copy.deepcopy(_config)
+
+    # Calculate total timesteps.
+    n_devices = len(jax.devices())
+    config.num_devices = n_devices
+    config = check_total_timesteps(config)
+    assert (
+        config.arch.num_updates > config.arch.num_evaluation
+    ), "Number of updates per evaluation must be less than total number of updates."
+
+    # Create the environments for train and eval.
+    env, eval_env = environments.make(config=config)
+
+    # PRNG keys.
+    key, key_e, q_net_key = jax.random.split(jax.random.PRNGKey(config.arch.seed), num=3)
+
+    # Setup learner.
+    learn, eval_q_network, learner_state = learner_setup(env, (key, q_net_key), config)
+
+    # Setup evaluator.
+    evaluator, absolute_metric_evaluator, (trained_params, eval_keys) = evaluator_setup(
+        eval_env=eval_env,
+        key_e=key_e,
+        eval_act_fn=get_distribution_act_fn(config, eval_q_network.apply),
```

Comment:
Now this might not work if noise_key is required during the evaluation however we need to partially discuss if this is required since i dont remember if they turn off the noise for evaluation.